### PR TITLE
Stop using CallAllUnordered in peer_set::add_initial_peers

### LINKED ISSUE #1696: Stop using CallAllUnordered in zebra_network::peer_set
**Is your feature request related to a problem? Please describe.**

`zebra-network::peer_set::initialize::add_initial_peers` uses `CallAllUnordered`, which misuses `poll_ready`.

If the connector uses a buffer in future, `CallAllUnordered` could fill it up, potentially causing hangs.

See #1593 for a general explanation.

**Describe the solution you'd like**

Use the `ServiceExt::oneshot` pattern in a loop, to avoid the `poll_ready` buffer slot leaks in `CallAllUnordered`.

PR #1620 contains a similar change in its `BlocksByHash` request:
```rust
               hashes
                    .into_iter()
                    .map(|hash| zs::Request::Block(hash.into()))
                    .map(|request| self.state.clone().oneshot(request))
                    .collect::<futures::stream::FuturesOrdered<_>>()
```
But we want a `FuturesUnordered` here.

https://github.com/ZcashFoundation/zebra/pull/1620/files#diff-3f14971a74a109913a4809bd1a77f4e0145e1edfa72aaded8f1fdcb2cc45f4f1L188

See #1593 for background.

**Describe alternatives you've considered**

Do nothing: if future changes add a buffer, the code might hang at startup, or fill up a bunch of buffer slots and make hangs more likely.

**Additional context**

This issue was discovered during the review in #1593. It is a bug fix for risky code.

### PR DESCRIPTION
**Pull request by oxarbitrage**
## Motivation

Stop using `CallAllUnordered` as suggested in https://github.com/ZcashFoundation/zebra/issues/1696

## Solution

Work in progress. Having some problems so looking for some hints. I will like @teor2345 to take a look if possible.

I am trying to collect in a vector the `tower::discover::Change<std::net::SocketAddr, peer::client::Client>` that in theory can be used as arguments to `tx.send()`.

However when i execute this code i always get the error arm of the match:

```
[zebra-network/src/peer_set/initialize.rs:215] e = Os {
    code: 101,
    kind: Other,
    message: "Network is unreachable",
}
```

If i change to `FuturesOrdered` i geta similar but different error:

```
[zebra-network/src/peer_set/initialize.rs:215] e = Os {
    code: 111,
    kind: ConnectionRefused,
    message: "Connection refused",
}
```

I tried other alternatives however by now i will like to know what could be wrong in this approach or if i need to change it please suggest. Thanks. 

## Review

@teor2345 

## Related Tickets

Closes #1696.

### CODE REVIEW

File: zebra-network/src/peer_set/initialize.rs [hunk 1]
```diff
@@ -194,15 +194,28 @@ where
...
-    while let Some(handshake_result) = handshakes.next().await {
-        tx.send(handshake_result).await?;
+    use futures::TryStreamExt;
+
+    let fut = initial_peers
+        .clone()
+        .into_iter()
+        .map(|request| connector.clone().oneshot(request))
+        .collect::<futures::stream::FuturesUnordered<_>>()
+        .try_collect::<Vec<_>>();
```
Comment (teor2345) [inline zebra-network/src/peer_set/initialize.rs]: We don't want to use `try_collect` here:

> If an error happens then all collected elements will be dropped and the error will be returned.

https://docs.rs/futures/0.3.12/futures/stream/trait.TryStreamExt.html#method.try_collect

Comment (teor2345) [reply zebra-network/src/peer_set/initialize.rs]: You can keep most of the existing code if you just drop the `try_collect`.

File: zebra-network/src/peer_set/initialize.rs [hunk 2]
```diff
@@ -194,9 +194,10 @@ where
     S::Future: Send + 'static,
 {
     info!(?initial_peers, "Connecting to initial peer set");
-    use tower::util::CallAllUnordered;
-    let addr_stream = futures::stream::iter(initial_peers.into_iter());
-    let mut handshakes = CallAllUnordered::new(connector, addr_stream);
+    let mut handshakes = initial_peers
+        .iter()
+        .map(|request| connector.clone().oneshot(*request))
+        .collect::<futures::stream::FuturesUnordered<_>>();
```
Comment (teor2345) [inline zebra-network/src/peer_set/initialize.rs:200]: We don't want to use `try_collect` here:
> If an error happens then all collected elements will be dropped and the error will be returned.

https://docs.rs/futures/0.3.12/futures/stream/trait.TryStreamExt.html#method.try_collect

Instead, we just pass the `FuturesUnordered` to the existing code, which expects a `CallAllUnordered` (they are similar types).

Comment (teor2345) [reply zebra-network/src/peer_set/initialize.rs:200]: I also removed a `clone` for you.

Comment (teor2345) [reply zebra-network/src/peer_set/initialize.rs:200]: Was the example code from PR #1620 helpful?
```rust
               hashes
                    .into_iter()
                    .map(|hash| zs::Request::Block(hash.into()))
                    .map(|request| self.state.clone().oneshot(request))
                    .collect::<futures::stream::FuturesOrdered<_>>()
```
https://github.com/ZcashFoundation/zebra/pull/1620/files#diff-3f14971a74a109913a4809bd1a77f4e0145e1edfa72aaded8f1fdcb2cc45f4f1L188

**Review by teor2345 — COMMENTED** [review]
I deleted a bunch of code, so @yaahc or @oxarbitrage should review these changes.

File: zebra-network/src/peer_set/initialize.rs (hunk 2 above)
Comment (oxarbitrage) [reply zebra-network/src/peer_set/initialize.rs:200]: Thanks @teor2345 , this was a lot simpler than i was expecting. Also i can now understand the cause of the error i was having.

**Review by oxarbitrage — COMMENTED** [review]
definitely looks good.

**Review by teor2345 — APPROVED** [review]
I think this is pretty straightforward and we should just merge it.

**Pull request merged**